commands/kubernetes.go

```diff
@@ -395,22 +405,169 @@ func RunKubernetesKubeconfigShow(c *CmdConfig) error {
 	return err
 }
 
+func cachedExecCredentialPath(id string) string {
+	return filepath.Join(kubeconfigCachePath(), id+".json")
+}
+
+// loadCachedExecCredential attempts to load the cached exec credential from disk. Never errors
+// Returns nil if there's no credential, if it failed to load it, or if it's expired.
+func loadCachedExecCredential(id string) *clientauthentication.ExecCredential {
+	path := cachedExecCredentialPath(id)
+	f, err := os.Open(path)
+	if err != nil {
+		return nil
+	}
+
+	defer f.Close()
+
+	var execCredential *clientauthentication.ExecCredential
+	if err := json.NewDecoder(f).Decode(&execCredential); err != nil {
+		if Verbose {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		return nil
+	}
+
+	if execCredential.Status == nil {
+		return nil
+	}
+
+	t := execCredential.Status.ExpirationTimestamp
+	if t.IsZero() || t.Time.Before(time.Now()) {
+		if err := os.Remove(path); err != nil && Verbose {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		return nil
+	}
+
+	return execCredential
+}
+
+// cacheExecCredential attempts to cache an ExecCredential. Never errors.
+func cacheExecCredential(id string, execCredential *clientauthentication.ExecCredential) {
+	// Don't bother caching if there's no expiration set
+	if execCredential.Status.ExpirationTimestamp.IsZero() {
+		return
+	}
+
+	cachePath := kubeconfigCachePath()
+	if err := os.MkdirAll(cachePath, os.FileMode(0700)); err != nil {
+		if Verbose {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		return
+	}
+
+	path := cachedExecCredentialPath(id)
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_TRUNC, os.FileMode(0600))
+	if err != nil {
+		if Verbose {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		return
+	}
+	defer f.Close()
+
+	if err := json.NewEncoder(f).Encode(execCredential); err != nil && Verbose {
+		fmt.Fprintln(os.Stderr, err)
+	}
+}
+
+// RunKubernetesKubeconfigExecCredential displays the exec credential. It is for internal use only.
+func RunKubernetesKubeconfigExecCredential(c *CmdConfig) error {
```

Comment:
Why is this exported if it's for internal use only? I would unexport it.

Reply:
The CLI command is for internal use, the function is exported because the others are too. I don't feel strongly either way.